File: compat/suites/python-sdk/groups/sts.py
"""
groups/sts.py — STS compatibility test implementations for the Python suite.
"""

from __future__ import annotations
import base64
import json
import time
from lib.harness import TestContext
from lib.clients import make_clients


def _sts(ctx: TestContext):
    return make_clients(ctx.endpoint, ctx.region).sts


# ── sts-identity ──────────────────────────────────────────────────────────────

def GetCallerIdentity(ctx: TestContext) -> None:
    sts = _sts(ctx)
    resp = sts.get_caller_identity()
    if not resp.get("Account"):
        raise AssertionError("GetCallerIdentity: missing Account")
    if not resp.get("UserId"):
        raise AssertionError("GetCallerIdentity: missing UserId")
    if not resp.get("Arn"):
        raise AssertionError("GetCallerIdentity: missing Arn")


def GetSessionToken(ctx: TestContext) -> None:
    sts = _sts(ctx)
    resp = sts.get_session_token(DurationSeconds=900)
    creds = resp.get("Credentials", {})
    if not creds.get("AccessKeyId"):
        raise AssertionError("GetSessionToken: missing AccessKeyId")
    if not creds.get("SecretAccessKey"):
        raise AssertionError("GetSessionToken: missing SecretAccessKey")
    if not creds.get("SessionToken"):
        raise AssertionError("GetSessionToken: missing SessionToken")


def GetFederationToken(ctx: TestContext) -> None:
    sts = _sts(ctx)
    resp = sts.get_federation_token(
        Name="compat-test",
        DurationSeconds=900,
        Policy=json.dumps({
            "Version": "2012-10-17",
            "Statement": [{"Effect": "Allow", "Action": "s3:GetObject", "Resource": "*"}],
        }),
    )
    creds = resp.get("Credentials", {})
    if not creds.get("AccessKeyId"):
        raise AssertionError("GetFederationToken: missing AccessKeyId")


# ── sts-assume ────────────────────────────────────────────────────────────────

def AssumeRole(ctx: TestContext) -> None:
    sts = _sts(ctx)
    resp = sts.assume_role(
        RoleArn=f"arn:aws:iam::000000000000:role/{ctx.run_id}-role",
        RoleSessionName=f"compat-{ctx.run_id}",
        DurationSeconds=900,
    )
    creds = resp.get("Credentials", {})
    if not creds.get("AccessKeyId"):
        raise AssertionError("AssumeRole: missing AccessKeyId")
    assumed = resp.get("AssumedRoleUser", {})
    if not assumed.get("AssumedRoleId"):
        raise AssertionError("AssumeRole: missing AssumedRoleId")


def AssumeRoleWithWebIdentity(ctx: TestContext) -> None:
    sts = _sts(ctx)
    # A syntactically valid but fake JWT (three base64url segments)
    header = base64.urlsafe_b64encode(json.dumps({"alg": "RS256"}).encode()).rstrip(b"=")
    payload = base64.urlsafe_b64encode(json.dumps({
        "sub": "test-user", "iss": "https://example.com", "exp": 9999999999,
    }).encode()).rstrip(b"=")
    fake_token = f"{header.decode()}.{payload.decode()}.fakesig"
    resp = sts.assume_role_with_web_identity(
        RoleArn=f"arn:aws:iam::000000000000:role/{ctx.run_id}-web-role",
        RoleSessionName=f"compat-web-{ctx.run_id}",
        WebIdentityToken=fake_token,
    )
    if not resp.get("Credentials", {}).get("AccessKeyId"):
        raise AssertionError("AssumeRoleWithWebIdentity: missing AccessKeyId")


# ── ImplMap ───────────────────────────────────────────────────────────────────

IMPLS = {
    "GetCallerIdentity": GetCallerIdentity,
    "GetSessionToken": GetSessionToken,
    "GetFederationToken": GetFederationToken,
    "AssumeRole": AssumeRole,
    "AssumeRoleWithWebIdentity": AssumeRoleWithWebIdentity,
}

SETUP = {}
TEARDOWN = {}

File: compat/suites/node-js-sdk/src/groups/sts.ts
/**
 * groups/sts.ts — STS compatibility test groups for the node-js-sdk suite.
 *
 * Status: NOT implemented in Overcast. Tests expected to fail with 501.
 *
 * Groups:
 *   sts-identity   — GetCallerIdentity, GetFederationToken
 *   sts-assume     — AssumeRole, GetSessionToken
 */

import {
  GetCallerIdentityCommand,
  AssumeRoleCommand,
  AssumeRoleWithWebIdentityCommand,
  GetSessionTokenCommand,
  GetFederationTokenCommand,
} from "@aws-sdk/client-sts"
import { makeClients } from "../lib/clients.js"
import type { TestGroup } from "../lib/harness.js"
import * as assert from "node:assert/strict";

export function makeSTSGroups(suite: string): TestGroup[] {
  return [
    // ── sts-identity ────────────────────────────────────────────────────────
    {
      suite,
      service: "sts",
      name: "sts-identity",
      tests: [
        {
          name: "GetCallerIdentity",
          fn: async (ctx) => {
            const { sts } = makeClients(ctx)
            const resp = await sts.send(new GetCallerIdentityCommand({}))
            assert.ok(resp.Account, "GetCallerIdentity: missing Account");
            assert.ok(resp.UserId, "GetCallerIdentity: missing UserId");
            assert.ok(resp.Arn, "GetCallerIdentity: missing Arn");
          },
        },
        {
          name: "GetSessionToken",
          fn: async (ctx) => {
            const { sts } = makeClients(ctx)
            const resp = await sts.send(new GetSessionTokenCommand({ DurationSeconds: 900 }))
            assert.ok(resp.Credentials?.AccessKeyId, "GetSessionToken: missing AccessKeyId");
            assert.ok(resp.Credentials?.SecretAccessKey, "GetSessionToken: missing SecretAccessKey");
            assert.ok(resp.Credentials?.SessionToken, "GetSessionToken: missing SessionToken");
          },
        },
        {
          name: "GetFederationToken",
          fn: async (ctx) => {
            const { sts } = makeClients(ctx)
            const resp = await sts.send(
              new GetFederationTokenCommand({
                Name: "compat-test",
                DurationSeconds: 900,
                Policy: JSON.stringify({
                  Version: "2012-10-17",
                  Statement: [{ Effect: "Allow", Action: "s3:GetObject", Resource: "*" }],
                }),
              }),
            )
            assert.ok(resp.Credentials?.AccessKeyId, "GetFederationToken: missing AccessKeyId");
          },
        },
      ],
    },

    // ── sts-assume ──────────────────────────────────────────────────────────
    {
      suite,
      service: "sts",
      name: "sts-assume",
      tests: [
        {
          name: "AssumeRole",
          fn: async (ctx) => {
            const { sts } = makeClients(ctx)
            // In Overcast the IAM role doesn't need to exist; we just want the
            // STS assume-role API to return valid temporary credentials.
            const resp = await sts.send(
              new AssumeRoleCommand({
                RoleArn: `arn:aws:iam::000000000000:role/${ctx.runId}-role`,
                RoleSessionName: `compat-${ctx.runId}`,
                DurationSeconds: 900,
              }),
            )
            assert.ok(resp.Credentials?.AccessKeyId, "AssumeRole: missing AccessKeyId");
            assert.ok(resp.AssumedRoleUser?.AssumedRoleId, "AssumeRole: missing AssumedRoleId");
          },
        },
        {
          name: "AssumeRoleWithWebIdentity",
          fn: async (ctx) => {
            const { sts } = makeClients(ctx)
            // A syntactically valid but fake JWT (three base64 segments).
            const fakeToken = [
              Buffer.from(JSON.stringify({ alg: "RS256" })).toString("base64url"),
              Buffer.from(
                JSON.stringify({ sub: "test-user", iss: "https://example.com", exp: 9999999999 }),
              ).toString("base64url"),
              "fakesig",
            ].join(".")
            const resp = await sts.send(
              new AssumeRoleWithWebIdentityCommand({
                RoleArn: `arn:aws:iam::000000000000:role/${ctx.runId}-web-role`,
                RoleSessionName: `compat-web-${ctx.runId}`,
                WebIdentityToken: fakeToken,
              }),
            )
            assert.ok(resp.Credentials?.AccessKeyId, "AssumeRoleWithWebIdentity: missing AccessKeyId");
          },
        },
      ],
    },
  ]
}
